proposals/0000-backpack.rst

```diff
@@ -0,0 +1,1185 @@
+        long.  (GHC can extract the full unit identifier from the
+        installed library database, although there is generally no
+        need to do so at this point.)
+
+    2. ``-this-unit-id`` remains a unit identifier (since we need to
+       know how it is instantiated), but we also specify the
+       hashed unit identifier for this unit identifier using the
+       ``-this-hashed-unit-id`` flag.
+
+For example, given that ``<Str>`` is instantiated with
+``str-bytestring-0.2-xxx:Str``, we have::
+
+    ghc -this-unit-id "concat-indef-0.1-abcdefg[Str=str-bytestring-0.2-xxx:Str]" \
+        -this-hashed-unit-id "concat-indef-0.1-abcdefg-xyz12345"
+        --make Str.hsig Concat.hs
+
+    ghc -this-unit-id "stringutils-indef-0.1-xxx[Str=str-bytestring-0.2-xxx:Str]" \
+        -this-hashed-unit-id "stringutils-indef-0.1-xxx-hijklm"
+        -unit-id "concat-indef-0.1-abcdefg-xyz12345" \
+        --make Str.hsig StringUtils.hs
+
+Legacy (but implemented) behavior
+
+    A ``-this-unit-id`` now takes a fully instantiated ``UnitId``.  Note
+    that the unit identifiers from dependencies do NOT have the
+    substitution applied to them (GHC can read off the necesary
+    substitution from ``-this-unit-id``.)  Here it is::
+
+        ghc -this-unit-id "concat-indef-0.1-abcdefg[Str=str-bytestring-0.2-xxx:Str]" \
+            --make Str.hsig Concat.hs
+
+        ghc -this-unit-id "stringutils-indef-0.1-xxx[Str=str-bytestring-0.2-xxx:Str]" \
+            -unit-id "concat-indef-0.1-abcdefg[Str=<Str>]" \
+            --make Str.hsig StringUtils.hs
+
+Not all unit identifiers are valid as arguments to ``this-unit-id``:
+only fully generalized unit identifiers (in the case of typechecking
+uninstantiated mixed libraries) or unit identifiers with no free module
+variables (in the case of compiling an instantiated mixed library) are
+permissible.
+
+Installed library database
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The **installed library database** (previously known as the installed
+package database) records uninstantiated and instantiated libraries
+that have been typechecked or compiled (respectively) so that they can be reused
+in later invocations of GHC.
+
+Logically, the installed library database is composed of two parts:
+
+1. Uninstantiated libraries, which are uniquely
+   identified by component identifier.  These libraries have only been
+   typechecked, but they can be instantiated on the fly according to a
+   unit identifier.
+
+2. Fully instantiated libraries, which are uniquely identifed by
+   unit identifier with no free module variables in its instantiating
+   substitution. These libraries have been compiled.
+
+3. Non-Backpack libraries, which don't have any signatures and
+   thus can't be instantiated in any meaningful way.  These libraries
+   have been compiled.  These are uniquely identified by component
+   or unit identifier (as without instantiations, these identifiers
+   are equivalent.)
+
```

Comment:
For typechecking, you need to install the uninstantiated and definite libraries you depend on.

For compiling, you need the fully instantiated and definite libraries you depend on.